Features, read/write, restructure and update

rafaqz opened the pull request:
Apologies for this PR being a grab bag of changes. It was updated in conjunction with changes to GeoInterface.jl.

But the outcome is round-trip read/write for all objects to/from json, and hopefully feature parity with GeoJSON.jl. Files are also a little more organized now.

This needs https://github.com/JuliaGeo/GeoInterface.jl/pull/64 merged and version bump in GeoInterface.jl to pass.

rafaqz committed:
reforganise and add missing interface methods

rafaqz committed:
reorganise and test interface

rafaqz committed:
use FeatureTrait and FeatureCollectionTrait

rafaqz committed:
Use FeatureCollection instead of a Vector

rafaqz committed:
passing tests

rafaqz committed:
dont use GeoInterface branch in CI

rafaqz commented:
Ok @evetion and @visr this is ready to go. Once we merge I'll make a PR to GeoJSON.jl

visr commented:
Great, thanks, I'll go over it soon. Anything in particular that you're not sure about or that you want to discuss?

rafaqz commented:
Nothing in particular. Probably `write` is the only interesting part

visr committed:
run JuliaFormatter

visr committed:
=== nothing

visr committed:
remove double function definition

visr committed:
resolve ambiguity found by Aqua

```
coordinates(::GeoInterface.AbstractGeometryTrait, g::GeoJSONTables.Geometry) in GeoJSONTables at GeoJSONTables\src\geointerface.jl:3
coordinates(t::GeoInterface.AbstractPointTrait, geom) in GeoInterface at GeoInterface\J298z\src\fallbacks.jl:104

Possible fix, define
  coordinates(::GeoInterface.AbstractPointTrait, ::GeoJSONTables.Geometry)
```

visr committed:
fix and test reading GeoformatTypes.GeoJSON

visr commented:
Still have a few more pending changes, like addressing the test_broken (that was already there).

Do you mind if I make the CRS warnings an error that instructs the user to use a kwarg for non-RFC files?

visr committed:
indexing a FeatureCollection returns a Feature

visr committed:
always set crs to EPSG:4326

rafaqz commented on src/features.jl:
```diff
@@ -66,7 +66,7 @@ Base.IteratorEltype(::Type{<:FeatureCollection}) = Base.HasEltype()
 # read only AbstractVector
 Base.size(fc::FeatureCollection) = size(array(fc))
-Base.getindex(fc::FeatureCollection, i) = array(fc)[i]
+Base.getindex(fc::FeatureCollection, i) = Feature(array(fc)[i])
```
This is returning a type different to the `eltype` of the array, breaking the base julia array interface, and e.g. `broadcast`.

If you want it to return a `Feature`, then you will need to fix this line too:

```julia
struct FeatureCollection{O,A} <: AbstractVector{eltype(A)}
```

The eltype has to be `<: Feature`.

rafaqz replied:
This is the reason that test is still broken... I'm not sure it makes sense.

visr replied:
Yeah you're right, it's fixed here: https://github.com/visr/GeoJSONTables.jl/pull/14/commits/0eb72ae2728c28ae66cdddb4f0edabc71c5cf8f3#diff-d1e20e357675f88e1ae77acf2891272308c984b6b4775fc2932e9ede2c08f162R34.

That commit is rather big, I'd appreciate it if you had a look. Still have some more things I want to iron out but we're getting close.

rafaqz replied:
The elype is abstract now as `Feature` is parametric... This will propagate type instability. 

Having the eltype of `FeatureCollection` different to its first type parameter is also usually best avoided. 

We really need to get the actual Feature type at the point of FeatureCollection construction. But this all gets pretty messy, and is why I just avoided it and went with doing the conversion in GeoInterface.

Maybe profiling/benchmark array ops you want to do on FeatureCollection would clear up priorities. I guess returning the abstract type at least satisfies the array interface now.

visr replied:
Hmm yeah I'll look into it a bit more. Summing a property over 16k `Feature`s is 10x slower, compared to doing it over the `JSON3.Object`s.

```julia
@time sum(f -> f.SUB_AREA, fc)
# 18.695922 seconds (557.25 k allocations: 2.395 GiB, 2.00% gc time, 0.31% compilation time)
@time sum(f -> f.properties.SUB_AREA, GJ.features(fc))
# 1.991023 seconds (490.61 k allocations: 396.247 MiB, 5.79% gc time, 5.52% compilation time)
```

rafaqz replied:
Yeah I expected something like that. Also those allocations are pretty large.

Probably manually getting the type of the first `Feature` at construction time and using that as the first parameter of `FeatureCollection` and `AbstractVector` will get you back to the same speed as the raw json.

visr replied:
Your suggestion, together with putting the features back into the FeatureCollection as a field (to avoid many calls to `object.features`) solved the discrepancy entirely, see 982f41be658f8210619409c32b00c8eeffd471ee.

rafaqz replied:
So does that mean its not lazy now?

visr replied:
No it's still lazy. On read JSON3 constructs this semi-lazy JSON3.Object for the FeatureCollection. When we do object.feature that gives a JSON3.Array which is a view into the same lazy object. Nothing is materialized and nothing is copied.

rafaqz replied:
Ok awesome I think this is a good solution then. Thanks for fixing this

rafaqz replied:
I'm happy to merge this now if you are

visr replied:
I'll merge this PR into master, but there are still a few things I want to check / tidy up before putting this on top of GeoJSON.jl. I'll try to wrap it up soon and put it in a last grab-bag PR, ok?

rafaqz replied:
I can also organise the GeoJSON PR if you like. Would be good to merge that tomorrow if possible...

visr replied:
Yeah would be great if you could organise the GeoJSON PR. I'll limit my actions now mostly on things that would be breaking. I also noticed that for instance writing the large geojson mentioned above took 3 minutes. Perhaps that can be improved, but it wouldn't change the API, so I'll leave that for later.

visr committed:
support bbox for geometries

Requiring this led to quite the refactor. Now the geometry structs don't only store the `JSON3.Array` with their coordinates, but the entire object. This includes the optional bbox, and any GeoJSON object may have other members according to the spec. The object is not restricted to a `JSON3.Object`, since these are the products of reading JSON strings. If we want to create another sub-geometry using `GeoInterface.getgeom` we need to be able to create them as well, For this we use a NamedTuple in the geometry, similar to the `_lower` approach for writing.

visr committed:
FeatureCollection add type parameter T<:Feature and features field

The type parameter T is taken from peeking at the first feature if it exists. This is needed to make iterating over the features ini a featurecollection type stable.

The features field was added after I noticed that the change above hardly helped iteration performance. On every `iterate` call the features had to be taken out of the JSON3.Object, which cannot be inferred. So it is better to keep them in the object, since they are a view on the same tape anyway.
